peps/pep-9999.rst

```diff
@@ -0,0 +1,546 @@
+attributes read-only:
+
+.. code-block:: python
+
+    class Book:
+        id: ReadOnly[int]
+
+        def __init__(self, id: int, name: str) -> None:
+            self.id = id
+            self.name: ReadOnly = name
+
+The explicit type in ``ReadOnly[<type>]`` can be omitted if the declaration has
+an initializing value. A type checker should apply its usual type inference
+rules to determine the type of ``name``.
+
+If an attribute is already implied to be read-only, like in the frozen :ref:`classes`,
+explicit declarations should be permitted and seen as equivalent, except ``Final``
+additionally forbids overriding in subclasses:
+
+.. code-block:: python
+
+    @dataclass(frozen=True)
+    class Point:
+        x: ReadOnly[int]
+        y: Final[int]
+
+Initialization
+--------------
+
+Assignment to a ``ReadOnly`` attribute can only occur as a part of the declaration,
+or within ``__init__`` of the same class. There is no restriction to how many
+times the attribute can be assigned to.
+
+.. code-block:: python
+
+    from collections import abc
+    from typing import ReadOnly
+
+
+    class Band:
+        name: str
+        songs: ReadOnly[list[str]]
+
+        def __init__(self, name: str, songs: abc.Iterable[str] | None = None) -> None:
+            self.name = name
+            self.songs = []
+
+            if songs is not None:
+                # multiple assignments during initialization are fine
+                self.songs = list(songs)
+
+        def clear(self) -> None:
+            # error: assignment to read-only "songs" outside initialization
+            self.songs = []
+
+
+    band = Band(name="Bôa", songs=["Duvet"])
+    band.name = "Python"  # ok: "name" is not read-only
+    band.songs = []  # error: "songs" is read-only
+    band.songs.append("Twilight")  # ok: list is mutable
+
+
+    class SubBand(Band):
+        def __init__(self) -> None:
+            # error: cannot assign to a read-only attribute of base class
+            self.songs = []
+
+An initializing value at a class level can serve as a `flyweight <https://en.wikipedia.org/wiki/Flyweight_pattern>`_
+default for instances:
+
+.. code-block:: python
+
+    class Patient:
+        number: ReadOnly = 0
+
+        def __init__(self, number: int | None = None) -> None:
+            if number is not None:
+                self.number = number
+
+This feature should also be supported by ``Final`` attributes. Specifically,
+``Final`` attributes initialized in a class body **should no longer** imply ``ClassVar``,
+and should remain assignable to within ``__init__``.
+
+.. note::
+    Classes defining :data:`~object.__slots__` cannot make use of this feature.
+    An attribute with a class-level value cannot be included in slots,
+    effectively making it a class variable.
+    Type checkers may warn or suggest explicitly marking the attribute as a ``ClassVar``.
+
+Type checkers should warn on read-only attributes which may be left uninitialized
```

Comment:
```suggestion
Type checkers may warn on read-only attributes which may be left uninitialized
```
Making sure that attributes are initialized is not an easy problem. While Python has not really been trying to do it much, Typescript has been trying to do it for a long time. A search in TypeScript issues for "strictPropertyInitialization" yields 98 open issues and 540 closed issues.

TypeScript has an annotation to assert that something is initialized when the type checker can't see it, and Python doesn't, so Python is less-equipped to handle this.

I would like it if type checkers started trying to move towards this. But I think making it a requirement in this PEP is not the right way to push towards it.